---
title: "VIMCO: Variational Inference for Multiple Correlated Outcomes in Genome-wide Association Studies"
author: "Xingjie Shi"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

**vimco** package proves a Bayesian variable selection methods for GWAS data with multiple traits. Unlike in BVSR where each trait is analyzed seperately, **vimco** performs a joint analysis for the multiple traits, while accounting for correlation among the multiple traits.
 

## Installation
To install the development version of **vimco**, it's easiest to use the 'devtools' package. Note that REMI depends on the 'Rcpp' package, which also requires appropriate setting of Rtools and Xcode for Windows and Mac OS/X, respectively.

```{r, fig.show='hold', eval=FALSE}
library(devtools)
install_github("XingjieShi/VIMCO")
```

## Fit VIMCO for simulated data

### Simulate data
We first generate genotypes and multiple phenotypes based on the linear modle $\mathbf Y = \mathbf X \mathbf B + \mathbf E$ as follows:
```{r, fig.show='hold'}
library(mvtnorm)
n = 300
p = 200
K = 5
set.seed(20132014)
X   = rmvnorm(n, mean=rep(0, p))
sigma.beta = rep(1, K)
bet = matrix(0, nrow = p, ncol = K)
lambda = 0.05
eta = rbinom(p, 1, lambda)
alpha = 1
gam = matrix(rbinom(p*K, 1, alpha), ncol=K)
for (k in 1:K){
  bet[, k] = rnorm(p, mean = 0, sd = sigma.beta[k]) * gam[,k] * eta 
}
  
sigma = diag(rep(1, K))
lp  = X %*% bet
sigma.e = diag(sqrt(diag(var(lp)))) %*% sigma %*% diag(sqrt(diag(var(lp))))
err = rmvnorm(n, rep(0, K), sigma.e)
Y   = lp + err
```

### Fit BVSR
```{r, fig.show='hold'}
library(vimco)
fit_Ind = emInd(X, Y)
str(fit_Ind)
```

### Fit VIMCO
```{r, fig.show='hold'}
fit_Mul = emMultiple(X, Y)
str(fit_Mul)
```

### Fit VIMCO using initial values from BVSR
```{r, fig.show='hold'}
p = ncol(X)
mu0     = fit_Ind$mu
sigb0   = fit_Ind$sigb
Theta0  = matrix(0, nrow=ncol(Y), ncol=ncol(Y))
diag(Theta0)  =   1/c(fit_Ind$sige)
Lambda0 = rep(1, p)
Alpha0  = fit_Ind$Alpha 
lambda0 = 1
fit_Mul_Ini = emMultiple(X, Y, mu0, sigb0, Theta0, Lambda0, Alpha0, TRUE) 
str(fit_Mul_Ini)
```


## Fit VIMCO for GWAS data with multiple traits
To handle GWAS data that is in a _PLINK_ format (incudes sim.bed, sim.bim, sim.fam) directly, **vimco** package provides a function `vimco::vimco()`, check its help file for more details. Here we provide a simple examples. The following 3 _PLINK_ files are already provided in the packages: 

- sim.bed
- sim.bim
- sim.fam


```{r, fig.show='hold', eval=FALSE}
path <- system.file(package = "vimco")
setwd(path)
stringname <- "sim"
tmp <- vimco(stringname, nPheno = 4, fit = TRUE)
```


## Replicate simulation results in Shi et al. (2018)
All the simulation results can be reproduced by using the code at simulation. Before running simulation to reproduce the results, please familarize yourself with **vimco** using 'VIMCO' vignette. Simulation results can be reproduced by following steps:

- simulation.r This function can be run in a PC, it will output files, named h_0.1_g_0_rhoE_0.2_rhoX_0.8, which contain inference results of each replicate, for all mthods: VIMCO, BVSR, sLMM and mvLMM. 

- summary.r This function read all outputs from simulation.r and calculate Power, FDR and AUC. It will output all summary results in _rds_ format, e.g. h_0.1_g_0_rhoE_0.2_rhoX_0.8.rds.

- barplot.r This function produce all figures in Shi et al. (2018).



## Reference
[VIMCO: Variational Inference for Multiple Correlated Outcomes in Genome-wide Association Studies](https://arxiv.org/abs/1807.10467)
